### PR TITLE
refactor(ui): use React Router v6 Link components

### PR DESCRIPTION
**Pull request by huwshimi**
## Done

- Update React Router Links to the v6 components
- Update tests to support the v6 compatibility router.

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps

- Load the React client.
- Make sure you can navigate through the app via the header links and links within the content.

## Fixes

Fixes: #3958.

### CODE REVIEW

File: ui/src/app/App.test.tsx [hunk 1]
```diff
@@ -38,7 +39,9 @@ describe("App", () => {
     const wrapper = mount(
       <Provider store={store}>
         <MemoryRouter initialEntries={[{ pathname: "/settings" }]}>
-          <App />
+          <CompatRouter>
```
Comment (huwshimi) [inline ui/src/app/App.test.tsx:42]: This same change is made in all the test files containing links.

File: ui/src/app/base/components/DHCPTable/DHCPTable.tsx [hunk 2]
```diff
@@ -2,7 +2,7 @@ import { useEffect, useState } from "react";
 
 import { List, MainTable, Spinner } from "@canonical/react-components";
 import { useDispatch, useSelector } from "react-redux";
-import { Link } from "react-router-dom";
+import { Link } from "react-router-dom-v5-compat";
```
Comment (huwshimi) [inline ui/src/app/base/components/DHCPTable/DHCPTable.tsx:5]: These are the changes to use the v6 Links.

**Comment by anthonydillon** [comment]
Out of curiosity, what is the benefit of v6 over the older version of Link?

**Comment by huwshimi** [comment]
> Out of curiosity, what is the benefit of v6 over the older version of Link?

There's nothing specific that we need for `Link` it's just a step towards upgrading React Router itself to v6 to save a really gigantic PR (as if this PR wasn't gigantic enough!)

**Comment by huwshimi** [comment]
> > > Out of curiosity, what is the benefit of v6 over the older version of Link?
> > 
> > 
> > There's nothing specific that we need for `Link` it's just a step towards upgrading React Router itself to v6 to save a really gigantic PR (as if this PR wasn't gigantic enough!)
> 
> Yeah exactly, such a large diff for a dep change. Something seems overly complex in this system. Just an observation.

Maybe to give the full picture, the reason that this is necessary is that React Router introduced a bunch of breaking changes in v6.

To ease the transition they introduced the compatibility router so that you can use v5 and v6 alongside one another.

This PR introduces that compatibility layer as well as upgrading the Links, however the reason this PR is so large is that the compatibility layer needed to be introduced in all tests that involve routing (which is 99% of the changes in the PR).

Once we've updated everything to React Router v6 we can remove the compatibility router.

**Comment by huwshimi** [comment]
> Oh sorry, I understand that and it's not an opinion of complexity in this project but in React router.

Yeah it's definitely a little unfortunate that the upgrade is so complex. I think they realised that which is why they made the compatibility layer, but it's still a lot of work!

**Review by Caleb-Ellis — APPROVED** [review]
LGTM! It's a shame the upgrade is such a hassle but at least this gets us started on an incremental upgrade

**Pull request merged**